spec.md

```diff
@@ -1090,6 +1084,252 @@ message ControllerServiceCapability {
 
 If the plugin is unable to complete the ControllerGetCapabilities call successfully, it MUST return a non-ok gRPC code in the gRPC status.
 
+#### `CreateSnapshot`
+
+A Controller Plugin MUST implement this RPC call if it has `CREATE_DELETE_SNAPSHOT` controller capability.
+This RPC will be called by the CO to create a new snapshot from a source volume on behalf of a user.
+
+This operation MUST be idempotent.
+If a snapshot corresponding to the specified snapshot `name` already exists and is compatible with the specified `source_volume_id` and `parameters` in the `CreateSnapshotRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateSnapshotResponse`.
+
+Snapshot can be used as the source to provision a new volume.
+CreateVolumeRequest is modified to take an OPTIONAL parameter of a source snapshot.
+Reverting snapshot, where data in the original volume is erased and replaced with data in the snapshot, is an advanced functionality not every storage system can support and therefore is currently out of scope.
+
+```protobuf
+message CreateSnapshotRequest {
+  // The ID of the source volume to be snapshotted.
+  // This field is REQUIRED.
+  string source_volume_id = 1;
+
+  // The suggested name for the snapshot. This field is REQUIRED for
+  // idempotency.
+  string name = 2;
+
+  // Secrets required by plugin to complete snapshot creation request.
+  // This field is OPTIONAL.
+  map<string, string> create_snapshot_secrets = 3;
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  // This field is OPTIONAL. The Plugin is responsible for parsing and
+  // validating these parameters. COs will treat these as opaque.
+  // Use cases for opaque parameters:
+  // - Specify a policy to automatically clean up the snapshot.
+  // - Specify an expiration date for the snapshot.
+  // - Specify whether the snapshot is readonly or read/write.
+  // - Specify if the snapshot should be replicated to some place.
+  // - Specify primary or secondary for replication systems that
+  //   support snapshotting only on primary.
+  map<string, string> parameters = 4;
+}
+
+message CreateSnapshotResponse {
+  // Contains all attributes of the newly created snapshot that are
+  // relevant to the CO along with information required by the Plugin
+  // to uniquely identify the snapshot. This field is REQUIRED.
+  Snapshot snapshot = 1;
+}
+
+// The information about a provisioned snapshot.
+message Snapshot {
+  // SnapshotSize is the disk size of the snapshot. The purpose of this
```

Comment:
nit: `disk size of the snapshot` -> `disk size of the restored snapshot`?

Reply:
Actually I changed this to "disk size of the snapshot" based on the Google Cloud API doc.

From this doc (https://cloud.google.com/compute/docs/reference/rest/v1/disks/createSnapshot), the output of createSnapshot has two sizes:
snapshot.diskSizeGb - "Size of the snapshot in GB."
snapshot.storageBytes - "A size of the storage used by the snapshot. As snapshots share storage, this number is expected to change with snapshot creation/deletion."

From here (https://cloud.google.com/compute/docs/reference/rest/v1/disks), there is a sizeGb in create volume:
sizeGb - "Size of the persistent disk, specified in GB. You can specify this field when creating a persistent disk using the sourceSnapshot parameter. The value of sizeGb must not be less than the size of the snapshot."

So our SnapshotSize is the same as snapshot.diskSizeGb - "Size of the snapshot in GB".  I can change this just to "size of the snapshot".

Reply:
Sound good!  I changed it in the latest PR.